x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -0,0 +1,309 @@
+                InputType.SEARCH,
+                InferModelAction.Request.DEFAULT_TIMEOUT_FOR_API
+            );
+
+            queryRewriteContext.registerAsyncAction(
+                (client, listener) -> executeAsyncWithOrigin(
+                    client,
+                    ML_ORIGIN,
+                    InferenceAction.INSTANCE,
+                    inferenceRequest,
+                    listener.delegateFailureAndWrap((l, inferenceResponse) -> {
+                        inferenceResultsSupplier.set(inferenceResponse.getResults());
+                        l.onResponse(null);
+                    })
+                )
+            );
+        } else {
+            // The inference ID can be null if either the field name or index name(s) are invalid (or both).
+            // If this happens, we set the "no inference results" flag to true so the rewrite process can continue.
+            // Invalid index names will be handled in the transport layer, when the query is sent to the shard.
+            // Invalid field names will be handled when the query is re-written on the shard, where we have access to the index mappings.
+            noInferenceResults = true;
+        }
+
+        return new SemanticQueryBuilder(this, noInferenceResults ? null : inferenceResultsSupplier, null, noInferenceResults);
+    }
+
+    private static InferenceResults validateAndConvertInferenceResults(
+        SetOnce<InferenceServiceResults> inferenceResultsSupplier,
+        String fieldName
+    ) {
+        InferenceServiceResults inferenceServiceResults = inferenceResultsSupplier.get();
+        if (inferenceServiceResults == null) {
+            return null;
+        }
+
+        List<? extends InferenceResults> inferenceResultsList = inferenceServiceResults.transformToCoordinationFormat();
+        if (inferenceResultsList.isEmpty()) {
+            throw new IllegalArgumentException("No inference results retrieved for field [" + fieldName + "]");
+        } else if (inferenceResultsList.size() > 1) {
+            // The inference call should truncate if the query is too large.
+            // Thus, if we receive more than one inference result, it is a server-side error.
+            throw new IllegalStateException(inferenceResultsList.size() + " inference results retrieved for field [" + fieldName + "]");
+        }
+
+        InferenceResults inferenceResults = inferenceResultsList.get(0);
+        if (inferenceResults instanceof ErrorInferenceResults errorInferenceResults) {
+            throw new IllegalStateException(
+                "Field [" + fieldName + "] query inference error: " + errorInferenceResults.getException().getMessage(),
+                errorInferenceResults.getException()
+            );
+        } else if (inferenceResults instanceof WarningInferenceResults warningInferenceResults) {
+            throw new IllegalStateException("Field [" + fieldName + "] query inference warning: " + warningInferenceResults.getWarning());
```

Comment:
@davidkyle When would the inference service return `ErrorInferenceResults` or `WarningInferenceResults`? Should these result types be handled as user errors or server-side errors?